dotnet build fails, but exit code is still 0

Example code: https://github.com/aaron-tyler-ds/-MSB3245

We have a chunk of code where we're forcing several msbuild warnings to be promoted to errors. We're seeing that this inconsistently causes `dotnet build` to return a exit code of 0 even though the build failed. I've distilled an example down to the repo mentioned at the start of this issue.

Example output. Note that the MSB3245 hit shows up as an error, and the summary view correctly lists 1 error, but the process exit code was still 0.
```
C:\repos\MSB3277\src\exe>dotnet build
Microsoft (R) Build Engine version 16.7.0+7fb82e5b2 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  Restored C:\repos\MSB3277\src\classlib1\classlib1.csproj (in 124 ms).
  Restored C:\repos\MSB3277\src\classlib2\classlib2.csproj (in 124 ms).
  Restored C:\repos\MSB3277\src\classlib3\classlib3.csproj (in 124 ms).
  Restored C:\repos\MSB3277\src\exe\exe.csproj (in 124 ms).
  classlib1 -> C:\repos\MSB3277\src\classlib1\bin\Debug\net47\classlib1.dll
  classlib2 -> C:\repos\MSB3277\src\classlib2\bin\Debug\net47\classlib2.dll
  classlib3 -> C:\repos\MSB3277\src\classlib3\bin\Debug\net47\classlib3.dll
C:\Program Files\dotnet\sdk\3.1.403\Microsoft.Common.CurrentVersion.targets(2084,5): error MSB3245: Could not resolve this reference. Could not locate the assembly "FooBar.Baz". Check to make sure the assembly exists on disk. If this reference is required by your code, you may get compilation errors. [C:\repos\MSB3277\src\exe\exe.csproj]
  exe -> C:\repos\MSB3277\src\exe\bin\Debug\net47\exe.exe

Build succeeded.

C:\Program Files\dotnet\sdk\3.1.403\Microsoft.Common.CurrentVersion.targets(2084,5): error MSB3245: Could not resolve this reference. Could not locate the assembly "FooBar.Baz". Check to make sure the assembly exists on disk. If this reference is required by your code, you may get compilation errors. [C:\repos\MSB3277\src\exe\exe.csproj]
    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:01.06

C:\repos\MSB3277\src\exe>echo %errorlevel%
0
```
Interestingly, when this project doesn't have `ProjectReferences` to other projects this occurs much less frequently.

The nonzero exit code breaks our automated build process. We've also seen other versions of this, where the process exit code is 1 but there are no errors listed, however I don't have a consistent repro of this behavior but it feels related.

--- a/src/Build/BackEnd/BuildManager/BuildManager.cs
+++ b/src/Build/BackEnd/BuildManager/BuildManager.cs
@@ -693,8 +693,6 @@ namespace Microsoft.Build.Execution
             BuildSubmission submission = PendBuildRequest(requestData);
             BuildResult result = submission.Execute();
 
-            SetOverallResultIfWarningsAsErrors(result);
-
             return result;
         }
 
@@ -2084,8 +2082,6 @@ namespace Microsoft.Build.Execution
                 // If the submission has completed or never started, remove it.
                 if (submission.IsCompleted || submission.BuildRequest == null)
                 {
-                    SetOverallResultIfWarningsAsErrors(submission.BuildResult);
-
                     _buildSubmissions.Remove(submission.SubmissionId);
 
                     // Clear all cached SDKs for the submission
@@ -2343,25 +2339,6 @@ namespace Microsoft.Build.Execution
             }
 
             return castPacket;
-        }
-
-        /// <summary>
-        /// Sets the overall result of a build only if the user had specified /warnaserror and there were any errors.
-        /// This ensures the old behavior stays intact where builds could succeed even if a failure was logged.
-        /// </summary>
-        private void SetOverallResultIfWarningsAsErrors(BuildResult result)
-        {
-            if (result?.OverallResult == BuildResultCode.Success)
-            {
-                ILoggingService loggingService = ((IBuildComponentHost)this).LoggingService;
-
-                if (loggingService.HasBuildSubmissionLoggedErrors(result.SubmissionId))
-                {
-                    result.SetOverallResult(overallResult: false);
-
-                    _overallBuildSuccess = false;
-                }
-            }
         }
 
         /// <summary>

--- a/src/Build/BackEnd/BuildManager/BuildSubmission.cs
+++ b/src/Build/BackEnd/BuildManager/BuildSubmission.cs
@@ -198,6 +198,13 @@ namespace Microsoft.Build.Execution
                 bool hasCompleted = (Interlocked.Exchange(ref _completionInvoked, 1) == 1);
                 if (!hasCompleted)
                 {
+                    // Did this submission have warnings elevated to errors? If so, mark it as
+                    // failed even though it succeeded (with warnings--but they're errors).
+                    if (((IBuildComponentHost)BuildManager).LoggingService.HasBuildSubmissionLoggedErrors(BuildResult.SubmissionId))
+                    {
+                        BuildResult.SetOverallResult(overallResult: false);
+                    }
+
                     _completionEvent.Set();
 
                     if (_completionCallback != null)

--- a/src/MSBuild.UnitTests/XMake_Tests.cs
+++ b/src/MSBuild.UnitTests/XMake_Tests.cs
@@ -2215,6 +2215,29 @@ namespace Microsoft.Build.UnitTests
             }
         }
 
+        [Fact]
+        public void EndToEndWarnAsErrors()
+        {
+            using TestEnvironment testEnvironment = UnitTests.TestEnvironment.Create();
+
+            string projectContents = ObjectModelHelpers.CleanupFileContents(@"<Project>
+
+  <Target Name=""IssueWarning"">
+    <Warning Text=""Warning!"" />
+  </Target>
+  
+</Project>");
+
+
+            TransientTestProjectWithFiles testProject = testEnvironment.CreateTestProjectWithFiles(projectContents, new string[0]);
+
+            bool success;
+
+            string output = RunnerUtilities.ExecMSBuild($"\"{testProject.ProjectFile}\" -warnaserror", out success, _output);
+
+            success.ShouldBeFalse(() => output);
+        }
+
 #if FEATURE_ASSEMBLYLOADCONTEXT
         /// <summary>
         /// Ensure that tasks get loaded into their own <see cref="System.Runtime.Loader.AssemblyLoadContext"/>.

--- a/src/MSBuild.UnitTests/XMake_Tests.cs
+++ b/src/MSBuild.UnitTests/XMake_Tests.cs
@@ -2230,7 +2230,7 @@ namespace Microsoft.Build.UnitTests
 
             TransientTestProjectWithFiles testProject = testEnvironment.CreateTestProjectWithFiles(projectContents);
 
-            string output = RunnerUtilities.ExecMSBuild($"\"{testProject.ProjectFile}\" -warnaserror", out bool success, _output);
+            RunnerUtilities.ExecMSBuild($"\"{testProject.ProjectFile}\" -warnaserror", out bool success, _output);
 
             success.ShouldBeFalse();
         }

--- a/src/MSBuild.UnitTests/XMake_Tests.cs
+++ b/src/MSBuild.UnitTests/XMake_Tests.cs
@@ -2228,14 +2228,11 @@ namespace Microsoft.Build.UnitTests
   
 </Project>");
 
+            TransientTestProjectWithFiles testProject = testEnvironment.CreateTestProjectWithFiles(projectContents);
 
-            TransientTestProjectWithFiles testProject = testEnvironment.CreateTestProjectWithFiles(projectContents, new string[0]);
+            string output = RunnerUtilities.ExecMSBuild($"\"{testProject.ProjectFile}\" -warnaserror", out bool success, _output);
 
-            bool success;
-
-            string output = RunnerUtilities.ExecMSBuild($"\"{testProject.ProjectFile}\" -warnaserror", out success, _output);
-
-            success.ShouldBeFalse(() => output);
+            success.ShouldBeFalse();
         }
 
 #if FEATURE_ASSEMBLYLOADCONTEXT

